File: cpp_merge/otc-agent/cpp_core/scripts/start-listeners.cpp
#include "start-listeners.hpp"
#include <future>
#include <iostream>
#include <stdexcept>

namespace elizaos {

std::future<void> main() {
    // NOTE: Auto-converted from TypeScript - may need refinement

    std::cout << "🚀 Starting token registration listeners...\n" << std::endl;

    // Check environment variables
    const auto requiredEnvVars = [;
    "NEXT_PUBLIC_REGISTRATION_HELPER_ADDRESS",
    "NEXT_PUBLIC_BASE_RPC_URL",
    "NEXT_PUBLIC_SOLANA_PROGRAM_ID",
    "NEXT_PUBLIC_SOLANA_RPC";
    ];

    const auto missingVars = requiredEnvVars.filter(varName => !process.env[varName]);

    if (missingVars.size() > 0) {
        std::cout << "⚠️  Missing environment variables:" << missingVars << std::endl;
        std::cout << "Some listeners may not start properly.\n" << std::endl;
    }

    // Start Base listener
    try {
        std::cout << "📡 Starting Base listener..." << std::endl;
        startBaseListener();
        std::cout << "✅ Base listener started\n" << std::endl;
        } catch (error) {
            std::cerr << "❌ Failed to start Base listener:" << error << std::endl;
        }

        // Start Solana listener
        try {
            std::cout << "📡 Starting Solana listener..." << std::endl;
            startSolanaListener();
            std::cout << "✅ Solana listener started\n" << std::endl;
            } catch (error) {
                std::cerr << "❌ Failed to start Solana listener:" << error << std::endl;
            }

            std::cout << "🎯 All listeners initialized successfully!" << std::endl;
            std::cout << "\n📝 Available endpoints:" << std::endl;
            std::cout << "  • POST /api/listeners/start - Start specific listeners" << std::endl;
            std::cout << "  • POST /api/listeners/backfill - Backfill historical events" << std::endl;
            std::cout << "\n💡 Use Ctrl+C to stop all listeners" << std::endl;

            // Keep the process running
            process.on[&]("SIGINT", () {
                std::cout << "\n👋 Shutting down listeners..." << std::endl;
                std::exit(0);
                });

                process.on[&]("SIGTERM", () {
                    std::cout << "\n👋 Shutting down listeners..." << std::endl;
                    std::exit(0);
                    });

}

} // namespace elizaos

File: otc-agent/scripts/start-listeners.ts
#!/usr/bin/env bun

/**
 * Listener Startup Script
 *
 * This script initializes blockchain event listeners for token registration.
 * It can be run manually or automatically during application startup.
 */

import { startBaseListener, backfillBaseEvents } from '../src/services/token-registration-listener-base';
import { startSolanaListener, backfillSolanaEvents } from '../src/services/token-registration-listener-solana';

async function main() {
  console.log('🚀 Starting token registration listeners...\n');

  // Check environment variables
  const requiredEnvVars = [
    'NEXT_PUBLIC_REGISTRATION_HELPER_ADDRESS',
    'NEXT_PUBLIC_BASE_RPC_URL',
    'NEXT_PUBLIC_SOLANA_PROGRAM_ID',
    'NEXT_PUBLIC_SOLANA_RPC'
  ];

  const missingVars = requiredEnvVars.filter(varName => !process.env[varName]);

  if (missingVars.length > 0) {
    console.warn('⚠️  Missing environment variables:', missingVars);
    console.log('Some listeners may not start properly.\n');
  }

  // Start Base listener
  try {
    console.log('📡 Starting Base listener...');
    await startBaseListener();
    console.log('✅ Base listener started\n');
  } catch (error) {
    console.error('❌ Failed to start Base listener:', error);
  }

  // Start Solana listener
  try {
    console.log('📡 Starting Solana listener...');
    await startSolanaListener();
    console.log('✅ Solana listener started\n');
  } catch (error) {
    console.error('❌ Failed to start Solana listener:', error);
  }

  console.log('🎯 All listeners initialized successfully!');
  console.log('\n📝 Available endpoints:');
  console.log('  • POST /api/listeners/start - Start specific listeners');
  console.log('  • POST /api/listeners/backfill - Backfill historical events');
  console.log('\n💡 Use Ctrl+C to stop all listeners');

  // Keep the process running
  process.on('SIGINT', () => {
    console.log('\n👋 Shutting down listeners...');
    process.exit(0);
  });

  process.on('SIGTERM', () => {
    console.log('\n👋 Shutting down listeners...');
    process.exit(0);
  });
}

// Run if called directly
if (import.meta.main) {
  main().catch(console.error);
}

export default main;